src/Cardano/Shell/DaedalusIPC.hs

```diff
@@ -0,0 +1,233 @@
+{-# LANGUAGE LambdaCase          #-}
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE RankNTypes          #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- |
+-- Copyright: © 2018-2019 IOHK
+--
+-- Provides a mechanism for Daedalus to discover what port the cardano-wallet
+-- server is listening on.
+--
+-- See <https://nodejs.org/api/child_process.html#child_process_child_process_spawn_command_args_options>
+-- for more information about the message protocol.
+
+module Cardano.Shell.DaedalusIPC
+    ( daedalusIPC
+    ) where
+
+import           Cardano.Prelude
+
+import           Cardano.BM.Trace (Trace, logError, logInfo, logNotice)
+import           Control.Concurrent (threadDelay)
+import           Control.Concurrent.Async (concurrently_, race)
+import           Control.Concurrent.MVar (MVar, newEmptyMVar, putMVar, takeMVar)
+import           Control.Exception (IOException, catch, tryJust)
+import           Control.Monad (forever)
+import           Data.Aeson (FromJSON (..), ToJSON (..), Value (..),
+                             eitherDecode, encode, object, withObject, (.:),
+                             (.=))
+import           Data.Bifunctor (first)
+import           Data.Binary.Get (getWord32le, getWord64le, runGet)
+import           Data.Binary.Put (putLazyByteString, putWord32le, putWord64le,
+                                  runPut)
+import           Data.Functor (($>))
+import           Data.Maybe (fromMaybe)
+import           Data.Text (Text)
+import           Data.Word (Word32, Word64)
+
+import           GHC.IO.Handle.FD (fdToHandle)
+import           System.Environment (lookupEnv)
+import           System.Info (arch)
+import           System.IO (Handle, hFlush, hGetLine, hSetNewlineMode,
+                            noNewlineTranslation)
+import           System.IO.Error (IOError, userError)
+import           Text.Read (readEither)
+
+import qualified Data.ByteString.Lazy as BL
+import qualified Data.ByteString.Lazy.Char8 as L8
+import qualified Data.Text as T
+
+----------------------------------------------------------------------------
+-- Daedalus <-> Wallet child process port discovery protocol
+
+data MsgIn = QueryPort
+    deriving (Show, Eq)
+data MsgOut = Started | ReplyPort Int | ParseError Text
```

Comment:
We reduce these even further. I think the 4 messages could just be one:
```haskell
data MsgOut = ListenPort Int deriving (Show, Eq)
```
Daedalus would need to be modified though.